core/src/main/cpp/msg/MessageBuilder.h

```diff
@@ -519,9 +434,72 @@ class MessageBuilder : public std::enable_shared_from_this<MessageBuilder>
+     *        ID token is provided.
+     * @param peerstd::shared_ptr<tokens::UserIdToken> peer user ID token. May be null.
+     * @param peerServiceTokens initial set of peer service tokens.
+     *        May be null.
+     * @param keyExchangeData key exchange data. May be null.
+     * @throws MslException if a user ID token is not bound to its master
+     *         token.
+     */
+    void initializeMessageBuilder(
+            int64_t messageId,
+            std::shared_ptr<msg::MessageCapabilities> capabilities,
+            std::shared_ptr<tokens::MasterToken> masterToken,
+            std::shared_ptr<tokens::UserIdToken> userIdToken,
+            std::set<std::shared_ptr<tokens::ServiceToken>> serviceTokens,
+            std::shared_ptr<tokens::MasterToken> peerMasterToken,
+            std::shared_ptr<tokens::UserIdToken> peerUserIdToken,
+            std::set<std::shared_ptr<tokens::ServiceToken>> peerServiceTokens,
+            std::shared_ptr<keyx::KeyExchangeFactory::KeyExchangeData> keyExchangeData);
+    /**
+     * <p>Construct a new message header with the provided message data.</p>
+     *
+     * <p>Headers are encrypted and signed. If a master token is provided, it
+     * will be used for this purpose. Otherwise the crypto context appropriate
+     * for the entity authentication scheme will be used. N.B. Either the
+     * entity authentication data or the master token must be provided.</p>
+     *
+     * <p>Peer tokens are only processed if operating in peer-to-peer mode.</p>
+     *
+     * @param ctx MSL context.
+     * @param entityAuthData the entity authentication data. May be null if a
+     *        master token is provided.
+     * @param masterToken the master token. May be null if entity
+     *        authentication data is provided.
+     * @param headerData message header data container.
+     * @param peerData message header peer data container.
+     * @throws MslEncodingException if there is an error encoding the JSON
+     *         data.
+     * @throws MslCryptoException if there is an error encrypting or signing
+     *         the message.
+     * @throws MslMasterTokenException if the header master token is not
+     *         trusted and needs to be to accept this message header.
+     * @throws MslEntityAuthException if there is an error with the entity
+     *         authentication data.
+     */
+    std::shared_ptr<MessageHeader> createMessageHeader(
+            std::shared_ptr<util::MslContext> ctx,
+            std::shared_ptr<entityauth::EntityAuthenticationData> entityAuthData,
+            std::shared_ptr<tokens::MasterToken> masterToken,
+            std::shared_ptr<MessageHeader::HeaderData> headerData,
+            std::shared_ptr<MessageHeader::HeaderPeerData> peerData);
+
+protected:
     /** MSL context. */
-    std::shared_ptr<util::MslContext> ctx_;
+    const std::shared_ptr<util::MslContext> ctx_;
```

Comment:
I'm not sure `const` on a `shared_ptr` matters. @padolph

Reply:
@padolph convinced me of this, I'll take out the const.